wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceUpgrader.java

```diff
@@ -129,15 +140,46 @@ private void upgradeWorkspacesInConnections(NamespaceSummary namespace) {
         ConnectorDetail detail = discoverer.getSpecification(namespace.getName(), connection,
                                                              new SpecGenerationRequest(path, Collections.emptyMap()));
 
-        SampleSpec spec = new SampleSpec(
-          connection, connectorName, path,
-          detail.getRelatedPlugins().stream().map(plugin -> {
-            ArtifactSelectorConfig artifact = plugin.getArtifact();
-            Plugin pluginSpec = new Plugin(
-              plugin.getName(), plugin.getType(), plugin.getProperties(),
-              new Artifact(artifact.getName(), artifact.getVersion(), artifact.getScope()));
-            return new StageSpec(plugin.getSchema(), pluginSpec);
-          }).collect(Collectors.toSet()));
+        // special handle of the database type, the current db connector supports browsing from schema/table depending
+        // on the database type, here it is safe to just use the old way to generate spec since:
+        // 1. the old db connection might not contain database in connection string
+        // 2. it does not support schema, so db type which supports schema can easily get messed up due to the path
+        //    not set up correctly. For example, for mysql, the getspec call will return the source schema since it
+        //    mysql itself does not support schema. But SqlServer will not return source schema because it supports
+        //    schema and the original workspace does not have this information. To make sure, the path does not mess up,
+        //    the above path for db connection will always be "/".
+        // 3. Using the old properties will ensure the old behavior get reserved for the db connectors
+        Schema dbSchema = null;
+        Map<String, String> dbProperties = Collections.emptyMap();
+        if (connectionType.equals(ConnectionType.DATABASE)) {
+          SchemaConverter schemaConvertor = new SchemaConverter();
+          try {
+            dbSchema = sample.isEmpty() ? null :
+                         schemaConvertor.toSchema("record", AbstractDirectiveHandler.createUberRecord(sample));
```

Comment:
is this calculating the schema from the workspace sample? Isn't that data transformed by whatever directives exist for the workspace? If so, it won't be the db schema.

Reply:
The initial sample will be the original data read from the database table. The directives are applied each time on the /execute call on this original data. The workspace will never store the executed result to make sure the directive can be removed. 
The old UI has this logic to get the specification for the pipeline:
1. call /specification in DatabaseHandler to get the properties(without schema) back
2. call /schema with an empty directive to get the source schema
3. call /schema with the directive applied to get the wrangler schema

Reply:
ok sounds good.